File: src/main/java/codility/lessons/lesson2/OddOccurrencesInArray.java
package codility.lessons.lesson2;

import java.util.HashMap;

public class OddOccurrencesInArray {
    public static int solution(int[] A) {
        int result = 0;

        for (int i : A) {
            result ^= i;
        }
        return result;
    }

    public static <T> T solutionT(T[] A) {
        T result = null;
        HashMap<T, Boolean> map = new HashMap<>();
        for (T i : A) {
            if (map.containsKey(i)) {
                map.remove(i);
            } else {
                map.put(i, true);
            }
        }
        for (T key : map.keySet()) {
            System.out.println(key);
        }

        if (map.keySet().size() > 0) {
            result = map.keySet().iterator().next();
        }
        return result;
    }
}

File: src/main/java/codility/lessons/lesson2/OddOccurrencesInArray.kt
package codility.lessons.lesson2

import java.util.*

// https://app.codility.com/programmers/lessons/2-arrays/odd_occurrences_in_array/
object OddOccurrencesInArray {
    @JvmStatic
    fun solution(A: IntArray): Int {
        return A.reduce { res, i -> res.xor(i) }
    }

    @JvmStatic
    fun <T> solutionT(A: Array<T>): T? {
        var result: T? = null
        val map = HashMap<T, Boolean>()
        for (i in A) {
            if (map.containsKey(i)) {
                map.remove(i)
            } else {
                map[i] = true
            }
        }
        if (map.keys.size > 0) {
            result = map.keys.iterator().next()
        }
        return result
    }
}